File: src/repomap.py
"""
Repomap — карта репозитория для агента (как в Aider).
Сканирует проект, извлекает классы/функции, даёт модели контекст.
"""
from __future__ import annotations
import ast
from pathlib import Path
from .real_tools import ToolResult


class RepomapTool:
    def __init__(self, workspace: Path):
        self.workspace = workspace

    def scan(self, max_files: int = 50, save_to: str = "") -> ToolResult:
        """Карта всех Python файлов проекта — классы, функции, импорты."""
        try:
            lines = [f"🗺️  Карта репозитория: {self.workspace.name}\n"]
            py_files = [
                p for p in sorted(self.workspace.rglob("*.py"))
                if "venv" not in str(p) and "__pycache__" not in str(p)
            ][:max_files]

            total_classes, total_fns = 0, 0
            for path in py_files:
                rel = path.relative_to(self.workspace)
                try:
                    tree = ast.parse(path.read_text(encoding="utf-8", errors="ignore"))
                except SyntaxError:
                    continue

                classes, fns = [], []
                for node in ast.walk(tree):
                    if isinstance(node, ast.ClassDef):
                        methods = [n.name for n in ast.walk(node) if isinstance(n, ast.FunctionDef)][:5]
                        classes.append(f"  📦 {node.name}({', '.join(methods)})")
                        total_classes += 1
                    elif isinstance(node, ast.FunctionDef) and not any(
                        isinstance(p, ast.ClassDef) for p in ast.walk(tree)
                        if hasattr(p, 'body') and node in getattr(p, 'body', [])
                    ):
                        fns.append(f"  ƒ {node.name}")
                        total_fns += 1

                if classes or fns:
                    lines.append(f"📄 {rel}")
                    lines.extend(classes[:3])
                    lines.extend(fns[:5])
                    lines.append("")

            lines.append(f"\n📊 Итого: {len(py_files)} файлов, {total_classes} классов, {total_fns} функций")
            output = "\n".join(lines)

            if save_to:
                Path(save_to).parent.mkdir(parents=True, exist_ok=True)
                Path(save_to).write_text(output, encoding="utf-8")
                output += f"\n💾 Сохранено: {save_to}"

            return ToolResult(success=True, output=output)
        except Exception as e:
            return ToolResult(success=False, output="", error=str(e))

    def file_summary(self, file_path: str) -> ToolResult:
        """Краткое содержание одного файла."""
        try:
            path = Path(file_path)
            if not path.exists():
                return ToolResult(success=False, output="", error=f"Файл не найден: {file_path}")
            tree = ast.parse(path.read_text(encoding="utf-8"))
            lines = [f"📄 {path.name}\n"]
            for node in ast.walk(tree):
                if isinstance(node, ast.ClassDef):
                    lines.append(f"  📦 class {node.name}:")
                    for item in node.body:
                        if isinstance(item, ast.FunctionDef):
                            args = [a.arg for a in item.args.args if a.arg != "self"]
                            lines.append(f"      def {item.name}({', '.join(args)})")
                elif isinstance(node, ast.FunctionDef):
                    args = [a.arg for a in node.args.args]
                    lines.append(f"  ƒ def {node.name}({', '.join(args)})")
            return ToolResult(success=True, output="\n".join(lines))
        except Exception as e:
            return ToolResult(success=False, output="", error=str(e))

File: src/repomap.rs
//! Repomap — карта репозитория для LLM (как в Aider). Порт `repomap.py`
//! (итерация 6). Оригинал парсит только `*.py` через модуль `ast`
//! (настоящий AST-парсер Python). Здесь эквивалента `ast` для
//! произвольного языка в чистом Rust нет (нужен был бы tree-sitter —
//! тяжёлая зависимость, которой нет в дереве), поэтому классы/функции
//! находятся простыми построчными regex — как и `diagram_tools.rs`
//! (`extract_classes` там) в этом же дереве делает для того же класса
//! задач. Расширено по сравнению с оригиналом: сканирует не только
//! `*.py`, а `.py/.rs/.go/.js/.ts/.java/.rb`.

use regex::Regex;
use std::path::Path;
use walkdir::WalkDir;

use crate::tools::ToolResult;

fn class_and_fn_patterns(ext: &str) -> Option<(Regex, Regex)> {
    match ext {
        "py" => Some((
            Regex::new(r"(?m)^class\s+(\w+)").unwrap(),
            Regex::new(r"(?m)^def\s+(\w+)\s*\(").unwrap(),
        )),
        "rs" => Some((
            Regex::new(r"(?m)^\s*(?:pub\s+)?(?:struct|enum|trait)\s+(\w+)").unwrap(),
            Regex::new(r"(?m)^\s*(?:pub\s+)?(?:async\s+)?fn\s+(\w+)").unwrap(),
        )),
        "go" => Some((
            Regex::new(r"(?m)^type\s+(\w+)\s+struct").unwrap(),
            Regex::new(r"(?m)^func\s+(?:\([^)]*\)\s*)?(\w+)\s*\(").unwrap(),
        )),
        "js" | "ts" | "jsx" | "tsx" => Some((
            Regex::new(r"(?m)^\s*(?:export\s+)?class\s+(\w+)").unwrap(),
            Regex::new(r"(?m)^\s*(?:export\s+)?(?:async\s+)?function\s+(\w+)").unwrap(),
        )),
        "java" | "kt" => Some((
            Regex::new(r"(?m)^\s*(?:public|private|protected)?\s*(?:class|interface)\s+(\w+)").unwrap(),
            Regex::new(r"(?m)^\s*(?:public|private|protected)?\s*(?:static\s+)?\w+\s+(\w+)\s*\([^)]*\)\s*\{").unwrap(),
        )),
        "rb" => Some((
            Regex::new(r"(?m)^\s*class\s+(\w+)").unwrap(),
            Regex::new(r"(?m)^\s*def\s+(\w+)").unwrap(),
        )),
        _ => None,
    }
}

pub fn scan(workspace: &Path, max_files: usize, save_to: Option<&str>) -> ToolResult {
    let mut lines = vec![format!(
        "🗺️  Карта репозитория: {}\n",
        workspace.file_name().and_then(|n| n.to_str()).unwrap_or(".")
    )];

    let source_exts = ["py", "rs", "go", "js", "ts", "jsx", "tsx", "java", "kt", "rb"];
    let mut files: Vec<_> = WalkDir::new(workspace)
        .into_iter()
        .filter_map(|e| e.ok())
        .filter(|e| e.file_type().is_file())
        .filter(|e| {
            let p = e.path().to_string_lossy();
            !p.contains("/venv/")
                && !p.contains("/__pycache__/")
                && !p.contains("/node_modules/")
                && !p.contains("/target/")
                && !p.contains("/.git/")
        })
        .filter(|e| {
            e.path()
                .extension()
                .and_then(|s| s.to_str())
                .map(|ext| source_exts.contains(&ext))
                .unwrap_or(false)
        })
        .collect();
    files.sort_by_key(|e| e.path().to_path_buf());
    files.truncate(max_files);

    let mut total_classes = 0;
    let mut total_fns = 0;
    let files_scanned = files.len();

    for entry in &files {
        let path = entry.path();
        let ext = path.extension().and_then(|s| s.to_str()).unwrap_or("");
        let Some((class_re, fn_re)) = class_and_fn_patterns(ext) else {
            continue;
        };
        let content = match std::fs::read_to_string(path) {
            Ok(c) => c,
            Err(_) => continue,
        };
        let rel = path.strip_prefix(workspace).unwrap_or(path);

        let classes: Vec<String> = class_re
            .captures_iter(&content)
            .map(|c| format!("  📦 {}", &c[1]))
            .collect();
        let fns: Vec<String> = fn_re
            .captures_iter(&content)
            .map(|c| format!("  ƒ {}", &c[1]))
            .collect();

        total_classes += classes.len();
        total_fns += fns.len();

        if !classes.is_empty() || !fns.is_empty() {
            lines.push(format!("📄 {}", rel.display()));
            lines.extend(classes.into_iter().take(3));
            lines.extend(fns.into_iter().take(5));
            lines.push(String::new());
        }
    }

    lines.push(format!(
        "\n📊 Итого: {} файлов, {} классов, {} функций",
        files_scanned, total_classes, total_fns
    ));
    let mut output = lines.join("\n");

    if let Some(save_path) = save_to {
        if let Some(parent) = Path::new(save_path).parent() {
            let _ = std::fs::create_dir_all(parent);
        }
        if std::fs::write(save_path, &output).is_ok() {
            output.push_str(&format!("\n💾 Сохранено: {}", save_path));
        }
    }

    ToolResult::success(output)
}

pub fn file_summary(file_path: &str) -> ToolResult {
    let path = Path::new(file_path);
    if !path.exists() {
        return ToolResult::error(format!("Файл не найден: {}", file_path));
    }
    let ext = path.extension().and_then(|s| s.to_str()).unwrap_or("");
    let Some((class_re, fn_re)) = class_and_fn_patterns(ext) else {
        return ToolResult::error(format!("Неподдерживаемое расширение файла: .{}", ext));
    };
    let content = match std::fs::read_to_string(path) {
        Ok(c) => c,
        Err(e) => return ToolResult::error(format!("Не удалось прочитать файл: {}", e)),
    };

    let mut lines = vec![format!("📄 {}\n", path.file_name().and_then(|n| n.to_str()).unwrap_or(file_path))];
    for cap in class_re.captures_iter(&content) {
        lines.push(format!("  📦 class {}", &cap[1]));
    }
    for cap in fn_re.captures_iter(&content) {
        lines.push(format!("  ƒ def {}", &cap[1]));
    }
    ToolResult::success(lines.join("\n"))
}
